src/sink/kafka.rs

```diff
@@ -0,0 +1,244 @@
+                    let payload = message.payload();
+                    let key = message.key();
+                    if payload.is_some() && key.is_some() {
+                        Some(self.try_payload(payload.unwrap(), key.unwrap()))
+                    } else {
+                        error!("Unable to retry message. It was lost to the ether.");
+                        KAFKA_PUBLISH_RETRY_FAILURE_SUM
+                            .fetch_add(1, Ordering::Relaxed);
+                        None
+                    }
+                })
+                .filter(|x| x.is_some())
+                .map(|x| x.unwrap())
+                .collect();
+            self.messages = new_messages;
+        }
+        self.message_bytes = 0;
+    }
+
+    fn flush_interval(&self) -> Option<u64> {
+        Some(self.flush_interval)
+    }
+
+    fn shutdown(mut self) -> () {
+        self.flush();
+    }
+}
+
+impl Kafka {
+    /// Send a payload to Kafka and return a future that will resolve to its
+    /// delivery result.
+    fn try_payload(&self, payload: &[u8], key: &[u8]) -> DeliveryFuture {
+        self.producer.send_copy(
+            &self.topic_name[..],
+            /* partition */ None,
+            Some(&payload[..]),
+            Some(&key[..]),
+            /* timestamp */ None,
+            /* block_ms */ 0,
+        )
+    }
+
+    /// Wait on all in-flight messages, and return an `OwnedMessage` for each message
+    /// that needs to be retried.
+    fn await_inflight_messages(&mut self) -> Vec<OwnedMessage> {
+        self.messages
+            .iter_mut()
+            .map(|future| {
+                let result = future.wait();
+                match result {
+                    Ok(inner) => match inner {
+                        Ok((_partition, _offset)) => {
+                            KAFKA_PUBLISH_SUCCESS_SUM.fetch_add(1, Ordering::Relaxed);
+                            None
+                        }
+
+                        Err((err, message)) => match err {
+                            KafkaError::MessageProduction(err) => match err {
+                                RDKafkaError::InvalidMessage
+                                | RDKafkaError::UnknownTopicOrPartition
+                                | RDKafkaError::LeaderNotAvailable
+                                | RDKafkaError::NotLeaderForPartition
+                                | RDKafkaError::RequestTimedOut
+                                | RDKafkaError::NetworkException
+                                | RDKafkaError::GroupLoadInProgress
+                                | RDKafkaError::GroupCoordinatorNotAvailable
+                                | RDKafkaError::NotCoordinatorForGroup
+                                | RDKafkaError::NotEnoughReplicas
+                                | RDKafkaError::NotEnoughReplicasAfterAppend
+                                | RDKafkaError::NotController => {
+                                    KAFKA_PUBLISH_RETRY_SUM
```

Comment:
You should kick out a log message detailing the error, possibly even telemetry. TBH it was a real pain to kick out telemetry for Elasticsearch error variants but it's come in handy in practice.

Reply:
After discussing this in an IM, it turns out there are a _ton_ of possible options here. I am pro doing a sum-up telemetry plus logs for added detail.